cql3/expr/expression.hh

```diff
@@ -183,14 +191,53 @@ struct function_call {
 
 struct cast {
     nested_expression arg;
-    cql3_type type;
+    std::variant<cql3_type, shared_ptr<cql3_type::raw>> type;
 };
 
 struct field_selection {
     nested_expression structure;
     shared_ptr<column_identifier_raw> field;
 };
 
+struct null {
+};
+
+struct bind_variable {
+    struct scalar {};
+    struct scalar_in {};
+    struct tuple {};
+    struct tuple_in {};
+    // FIXME: infer shape from expression rather than from grammar
+    std::variant<scalar, scalar_in, tuple, tuple_in> shape;
+    int32_t bind_index;
+};
+
+// A constant which does not yet have a date type. It is partially typed
+// (we know if it's floating or int) but not sized.
+struct untyped_constant {
+    enum type_class { integer, floating_point, string, boolean, duration, uuid, hex };
+    type_class partial_type;
+    sstring raw_text;
+};
+
+// Denotes construction of a tuple from its elements, e.g.  ('a', ?, some_column) in CQL.
+struct tuple_constructor {
+    std::vector<expression> elements;
+};
+
+// Constructs a collection of same-typed elements
+struct collection_constructor {
+    enum class style_type { list, set, map };
+    style_type style;
+    std::vector<expression> elements;
+};
+
+// Constructs an object of a user-defined type
+struct usertype_constructor {
+    using elements_map_type = std::unordered_map<column_identifier, nested_expression>;
```

Comment:
It looks like `std::map` works with incomplete types and doesn't require `nested_expression`. The code would be simpler, but maybe slower (?)

Reply:
I think going with unordered_map is better since that's what user_types::literal used. Later we can consider switching to std::map. I don't think there's much difference, since user types don't have enough members to make the container complexity important.

OTOH I don't think we have ordering for column_identifier, so std::map won't work out of the box.